Speed up hide_existing=true when listing candidates

nphilipp opened the pull request:
Previously, the database was queried for each build returned from koji, which could result in thousands of individual queries.

Instead, query the database once for known builds associated with an update and filter out returned builds found in this set. This set isn't filtered for active releases because a build might be associated with an update for an archived release but still inherited into one of the tags we query Koji for. It performs well enough at the moment but the set can easily be cached if not.

Fixes: #3687

------

This PR also reverts the respective workaround. To test the performance, configure your testing instance to query koji.stg instead of using the mocked instance, e.g. by using `vagrant up --use-staging`  or having something like this in your `development.ini` (substitute your FAS user in `krb_principal`, also you might need to install `krb5-workstation` and acquire a matching ticket in the Vagrant box):

````
[app:main]
buildsystem = koji
koji_hub = https://koji.stg.fedoraproject.org/kojihub
koji_web_url = https://koji.stg.fedoraproject.org/koji/
krb_principal = nphilipp@STG.FEDORAPROJECT.ORG
acl_system = pagure
pagure_url = https://src.stg.fedoraproject.org/
...
````

nphilipp committed:
Speed up hide_existing=true when listing candidates

Previously, the database was queried for each build returned from koji,
which could result in thousands of individual queries.

Instead, query the database once for known builds associated with an
update and filter out returned builds found in this set. This set isn't
filtered for active releases because a build might be associated with an
update for an archived release but still inherited into one of the tags
we query Koji for. It performs well enough at the moment but the set can
easily be cached if not.

Additionally, revert the workaround we put in place (commit
7a83faf579bc45fcf524bea87c6afc26d8a628f6) and test the feature also
for builds/tags of an archived release.

Fixes: #3687

Signed-off-by: Nils Philippsen <nils@redhat.com>

abompard commented:
This looks sane. There's a minor flake8 error, and also I think it'd be useful to add a test to cover the case you describe, where a build in an archived release is inherited, so that we don't accidentally change that later (like, when looking for more optimizations... ;-) )

nphilipp committed:
Allow idiomatic SQLA 'IS NOT NULL' comparisons

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp commented:
> I think it'd be useful to add a test...

Agreed.

nphilipp committed:
make robust against missing data in tag_releases

I only encountered crashes concerning this in testing with mocked data,
but better safe than sorry.

Signed-off-by: Nils Philippsen <nils@redhat.com>

abompard commented on setup.cfg:
```diff
@@ -15,7 +15,7 @@ width = 80
 max-line-length = 100
 exclude = .git,.tox,dist,*egg,build,tools
-ignore = E712,W503
+ignore = E711,E712,W503
```
Hmm I'm not sure we should disable it globally. Does SQLAlchemy accept the `is not None` construct? If it does we can use it, if not I think it's better to exclude that check on the specific line.
